### PR TITLE
lib/systems/parse.nix: add removeAbiSuffix, use it

### PR DESCRIPTION
**Pull request by ghost**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

`gnu-config` ignores the portion of a triple matching the regex `e?abi.*$`

In other words, `i386-linuxabichickenlips` is a valid triple.

This commit updates our parsing routines to match gnu-config.

I was recently surprised to discover that it is in fact possible to shoehorn ABI flavors into nix doubles in a way which preserves their property of being a (non-canonical) subset of the valid gnu-config triples.  This commit is required in order to exploit that discovery to add automatic detection of ILP32 platforms (64-bit void*, 32-bit int, like the Apple Watch and MIPS n32) to Nix.

### CODE REVIEW

**Review by alyssais — APPROVED** [review]

**Comment by Ericson2314** [comment]
> This commit is required in order to exploit that discovery to add automatic detection of ILP32 platforms (64-bit void*, 32-bit int, like the Apple Watch and MIPS n32) to Nix.

I thought GNU config was handling those in a different (also weird) way?

**Comment by Ericson2314** [comment]
Ah right, I am thinking of

- https://reviews.llvm.org/D125412
- https://reviews.llvm.org/D61259
- https://lists.gnu.org/archive/html/config-patches/2023-08/msg00059.html
- https://lists.gnu.org/archive/html/config-patches/2023-09/msg00003.html

which are...not exactly less nuts!

**Comment by amjoseph-nixpkgs** [comment]
If curious, the hack is:

```
$ config.sub mips64el-linuxabin32
mips64el-unknown-linux-gnuabin32
```

Since `mips64el-linuxabin32` has only one hyphen, Nix will accept it as a value for `builtins.currentSystem`.

**Comment by amjoseph-nixpkgs** [comment]
> https://reviews.llvm.org/D125412

This is so totally frustrating.

`gnu-config` had support for this [*six years ago*](https://wiki-archive.linaro.org/Platform/arm64-ilp32) and calls it `aarch64-*-*_ilp32`.  There's already support in GCC, binutils, and the Linux kernel ABI.

So when LLVM decides to add support do they use the same triple?  Noooo, they have to do their own thing apparently.  So they call it `arm64_32-*-*`.
